pylint/checkers/classes/class_checker.py

```diff
@@ -880,6 +887,39 @@ def visit_classdef(self, node: nodes.ClassDef) -> None:
         self._check_proper_bases(node)
         self._check_typing_final(node)
         self._check_consistent_mro(node)
+        self._check_declare_non_slot(node)
+
+    def _check_declare_non_slot(self, node: nodes.ClassDef) -> None:
+        if not self._has_valid_slots(node):
+            return
+
+        slot_names = self._get_classdef_slots_names(node)
+
+        for base in node.bases:
+            ancestor = safe_infer(base)
+            if not isinstance(ancestor, nodes.ClassDef):
+                continue
+            if not self._has_valid_slots(ancestor):
+                return
+            else:
+                slot_names.extend(self._get_classdef_slots_names(ancestor))
+
+        # Every class in bases has __slots__
+
+        if "__dict__" in slot_names:
+            return
+
+        for child in node.body:
+            if isinstance(child, nodes.AnnAssign):
+                if child.value is not None:
+                    continue
+                if isinstance(child.target, nodes.AssignName):
+                    if child.target.name not in slot_names:
+                        self.add_message(
+                            "declare-non-slot",
+                            args=child.target.name,
+                            node=child.target,
```

Comment:
could you add a confidence argument?

Reply:
I'm not certain how to know what confidence level I should use. but went with `confidence=INFERENCE` as this seems the standard for the class checker (and I have used inference in the check).

Reply:
Yes if you're using ``infer``, the confidence level is inference, otherwise HIGH. With the apparition of ruff, unless there's inference involved ruff can probably do it as well and faster than us.